Develop

Armyke opened the pull request:
Hi everybody,
I found your project really interesting. Having found the paper and the repository I thought it would be great to try an ML path for image alignment. Still had to adapt it a lot, I tried not to break compatibility. The only issue that there may be is the change of notation from "test" to "val" to leave space for an actual test_set, if it is intended to divide dataset in the classical 3 splits (train, val, test). As an added value I introduced:

-  TensorBoard logging
- pep8 formatting of the code
- fixed a bug that prevented using MSE loss
- some error handling and initialization of folders or paths as well as updated compatibility and fixed 
  deprecation problems for the new version of torch (1.3.0),
- adding a requirements.txt,
- learning rate schedulers and some parameters for training optimization.

In addition the repo is now compatible with another type of dataset (which I used) that is described in the updated README.md.
The next steps will be:
- fixing an issue which prevents graph logging in TB when using cuda, 
- log more interesting visualizations of training
- improve model exports (not only on each epoch) specifying the number of steps between each model validation and export
- adding the possibility of restarting training from a checkpoint.

Anyway the commits should be atomical and have clear messages, so you can cherry pick the ones you find most interesting/useful. Else, if needed, I also may be available for splitting this pull request into separate ones, each with a more specific purpose.
Hope I've been of help and looking forward to your reply!
Cheers

Commit:
Adding copy of synth dataset called coupled dataset yet to be implemented, as well as requirements

Commit:
Adding align_corners=True as parameter when generating grid to avoid stdout flooding of warnings of deprecation

Commit:
Huge pep8 refactoring, adding a main() func and a __main__ to run it, as well as separated fn to parse_flags

Commit:
Changed notation from test to val (validation), added variables train_csv_path and val_csv_path as the glob(*train.csv) and glob(*val.csv), removed unnecessary import

Commit:
Huge pep8 formatting, added numpy to requirements

Commit:
huge pep8 formatting

Commit:
Adding CoupleDataset possibility to train.py through an arg --coupled_dataset

Commit:
pep8 formatting

Commit:
Adding a Raise error when image is a strange format

Commit:
Formatting

Commit:
In coupled_dataset, adapted class to serve 2 images per , added a TODO.
In train if args.coupled_dataset adapted Dataset object creation.
In train_test_fn formatting.

Commit:
Adapted train.py to new type of dataset

Commit:
Formatting

Commit:
Adding preprocessing of image_a by blanking outside vertices in dataset serving pair (a, b) images

Commit:
Updated requirements.txt

Commit:
Removed unnecessary imports, refactoring + adding a new image pairs serving func

Commit:
Adapted methods to take 3 channels images and output the same dimensions

Commit:
Formatting

Commit:
Removed unnecessary argument

Commit:
Formatting + adding log_interval as parameter to specify or to accept default value

Commit:
Fixed bug where images where not rescaled to the right input dimension

Commit:
Changed all parser args to right notation (minus to underscore), added a check to trained_models_dir

Commit:
Adding tqdm in requirements, improved visualization of training with progress bar. TODO add accuracy!!!!

Commit:
Formatting of all py's that weren't still touched

Commit:
Updated doc_string

Commit:
Fixed opencv name in requirements.txt

Commit:
Added cosine scheduler and flag (default True) to use it or not

Commit:
Fixed bug that prevented usage of MSE loss

Commit:
Adding a flag for setting scheduler hparam T_max, adding a talking error when not finding csv train/val files

Commit:
Adding tensorboard logging of graph and loss

Commit:
Updated tb_dir with the trained_models filename flag

Commit:
Fixed compatibility bug in test (.item() instead of [0]), fixed bug that prevented computation if loss was MSE

Commit:
Wrapped shape of tensor in [] for readability

Commit:
Changed format of dataset csv to have only normalized bbs, adapted Dataset object

Commit:
Changed notation in test fn from training loss to test loss in tb logging

Commit:
Changed notation from test to val

Commit:
Updated README.md

Commit:
Updated README.md

Commit:
Updated README.md

Commit:
Changed notation from test to val also in already present csvs

Commit:
Changed notation from test to val

Commit:
Adding learning rate TB logging if a scheduler is defined

Commit:
Fixed value of the step in TB logging, didn't start from step 0

Commit:
Added doc string in parse_flags

Commit:
Updated gitignore

Commit:
Fixed bug that prevented evaluation and export

Commit:
Corrected checkpoint path to be in a folder parallel to logdir in the <trained_models_dir>/<trained_models_fn>/ folder

Commit:
Fixed compatibility bug: torch.autograd.variable.Variable doesn't exists anymore, substituted with torch.autograd.Variable

Commit:
minor fixes

Commit:
Merge branch 'develop' of https://github.com/Armyke/cnngeometric_pytorch into develop

ignacio-rocco commented:
Hi Michele, thanks for contribution!

I've tested the code and it seems to work nicely. I fixed some deprecation warnings and other small errors. There might still be some Variables() around (which are also deprecated) but it's not a big deal. Do you want to do some other fixes before I merge?

Armyke commented:
Hi Ignacio,
I'm very pleased to hear you so soon and I'm very happy that you like my work.
I'm still working a little bit on the repository, my next steps are pretty much the ones described in the pull request description.
I think it's fine to merge it now, due to the fact that now I am currently working on displaying insights of my model on TensorBoard and for what regards the training pipeline I think that there won't be any other major suggestions on my side.
I have some other minor ideas which we might well discuss on issues I may eventually open (e.g. specifying a number of steps after which validate the model and not only after each epoch).
Best regards

ignacio-rocco commented:
Hi @Armyke, I'm not sure I understand the use case of the CoupledDataset that you introduced. I see that you have pairs of training images, but I don't understand what's the role of theta in this case. Could you say a bit more about this?
Thanks!

Armyke commented:
Hi Ignacio,
the reason why I needed the possibility of a coupled dataset is to
supervisely train a SIFT-like alignment model.
What it does is that takes a couple of images which share an object where
vertices are annotated and aims to teach the model to warp vertices_a over
vertices_b and that's where theta matrix comes into play: it is in fact the
ground truth of this trasformation.
Of course is vital the sorting of the vertices which MUST be ordered
correctly depending on the orientation of the respective images.
For my purpose I used only affine transformations, which make necessary
only the first three vertices. The reason why all vertices should be
specified is that as preprocessing of image a, to make feature extraction
easier, image_a pixels are set to blank outside the vertices while image b
is served as is.
I can't assure the correct working of this mode when vertices are more than
4 cause wasn't part of my use case but I expect it not to have problems.
I hope it's all clear now, for every other question I'll be at disposition!
If you wish me to update readme I can work on it, just let me know.
Kind regards

Il Ven 20 Dic 2019, 16:24 Ignacio Rocco <notifications@github.com> ha
scritto:

> Hi @Armyke <https://github.com/Armyke>, I'm not sure I understand the use
> case of the CoupledDataset that you introduced. I see that you have pairs
> of training images, but I don't understand what's the role of theta in this
> case. Could you say a bit more about this?
> Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ignacio-rocco/cnngeometric_pytorch/pull/14?email_source=notifications&email_token=AKOSWP77GK3USXGLI2MJY43QZTPSVA5CNFSM4JPUJCD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHNGUVI#issuecomment-567962197>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKOSWP6MO6CXAZTRZ3A73STQZTPSVANCNFSM4JPUJCDQ>
> .
>

ignacio-rocco commented:
Hi Michele, I've updated the repo with some extra code from the extended TPAMI version of the paper. 

Given that your CoupledDataset seems quite specific for your task, and only works for the affine model,  I've removed it from the repo for now.

Best,
Ignacio

Armyke commented:
Hi Ignacio,
Sorry for the late response, busy days :)
Hope you still found something useful to held.
I'll definetely checkout the new paper as well as your new commits on the
repo.
Kind regards and happy new year!

Il Sab 21 Dic 2019, 17:51 Ignacio Rocco <notifications@github.com> ha
scritto:

> Hi Michele, I've updated the repo with some extra code from the extended
> TPAMI version of the paper.
>
> Given that your CoupledDataset seems quite specific for your task, and
> only works for the affine model, I've removed it from the repo for now.
>
> Best,
> Ignacio
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ignacio-rocco/cnngeometric_pytorch/pull/14?email_source=notifications&email_token=AKOSWPY5TZFYXDR73HODUGDQZZCRBA5CNFSM4JPUJCD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHO7RGA#issuecomment-568195224>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKOSWPZHUWJXNLA3CKD2XN3QZZCRBANCNFSM4JPUJCDQ>
> .
>
